### PR TITLE
toolkit: add chatbot response regeneration

### PR DESCRIPTION
**Pull request by danylo-boiko**
**Description**
- Added ability to regenerate chatbot responses

**AI Description**

This PR introduces a new feature for message regeneration, allowing users to regenerate the last message sent by a specific agent within a conversation. The changes are as follows:

- **src/backend/crud/message.py**: A new function, `delete_messages`, is added to delete messages by IDs. It takes a list of message IDs and a user ID as arguments and deletes the corresponding messages from the database.
- **src/backend/routers/chat.py**: The `process_message_regeneration` function is added to process message regeneration. It retrieves the last message sent by the specified agent, attaches files to messages, and returns a tuple containing necessary data to regenerate the message.
- **src/backend/services/chat.py**: The `process_message_regeneration` function is added to process message regeneration. It retrieves the last message sent by the specified agent, attaches files to messages, and returns a tuple containing necessary data to regenerate the message.
- **src/backend/tests/unit/crud/test_message.py**: A new test, `test_delete_messages`, is added to test the deletion of multiple messages by IDs. It creates three messages, deletes them, and verifies that they are no longer present in the database.
- **src/backend/tests/unit/routers/test_chat.py**: Two new tests, `test_stream_regeneration_existing_chat` and `test_stream_regeneration_existing_chat_without_user_messages`, are added to test the regeneration of chat streams for existing conversations with and without user messages.
- **src/interfaces/assistants_web/src/app/share/[id]/ReadOnlyConversation.tsx**: The `isReadOnly` prop is added to the `ReadOnlyConversation` component to indicate whether the conversation is read-only.
- **src/interfaces/assistants_web/src/assets/icons/Regenerate.tsx**: A new icon component, `Regenerate`, is added to represent the regeneration action. It imports React and SVG-related modules and exports a React functional component that renders an SVG path.
- **src/interfaces/assistants_web/src/assets/icons/index.ts**: The `Regenerate` icon is exported for use in other components.
- **src/interfaces/assistants_web/src/cohere-client/client.ts**: The `regenerate` property is added to the `CohereClient` class to enable message regeneration. The `getChatStreamEndpoint` method is modified to include the `/regenerate` endpoint when `regenerate` is true.
- **src/interfaces/assistants_web/src/components/Conversation/Conversation.tsx**: The `handleRegenerate` prop is added to the `Conversation` component to handle message regeneration.
- **src/interfaces/assistants_web/src/components/MessageRow/MessageRow.tsx**: The `isBotMessage` function is added to determine if a message is from a bot. The `isRegenerationEnabled` function is added to check if message regeneration is enabled. The `onRegenerate` prop is added to the `MessageRow` component to handle message regeneration.
- **src/interfaces/assistants_web/src/components/MessagingContainer/MessagingContainer.tsx**: The `onRegenerate` prop is added to the `MessagingContainer` component to handle message regeneration.
- **src/interfaces/assistants_web/src/components/UI/CopyToClipboardButton.tsx**: The `duration` and `hoverDelay` props are added to the `CopyToClipboardIconButton` component to control the duration and hover delay of the tooltip.
- **src/interfaces/assistants_web/src/components/UI/Icon.tsx**: The `Regenerate` icon is added to the `IconList` for use in the `getIcon` function.
- **src/interfaces/assistants_web/src/components/UI/Input.tsx**: The `duration` prop is added to the `Input` component to control the duration of the tooltip.
- **src/interfaces/assistants_web/src/hooks/use-chat.ts**: The `handleRegenerate` function is added to handle message regeneration. It finds the latest user message, checks if streaming is active, and regenerates the message if conditions are met.
- **src/interfaces/assistants_web/src/hooks/use-streamChat.ts**: The `regenerate` property is added to the `useStreamChat` hook to enable message regeneration.
- **src/interfaces/assistants_web/src/types/message.ts**: The `isBotMessage` function is added to determine if a message is from a bot.

### CODE REVIEW

**Comment by tianjing-li** [comment]
@danylo-boiko Thank you for this contribution! We will review it as soon as we can

File: src/backend/services/chat.py [hunk 1]
```diff
@@ -158,6 +159,134 @@ def process_chat(
...
+       Raises:
+           HTTPException: If there are no messages from the specified agent in the conversation.
+    """
+    agent_messages = [
+        message
+        for message in conversation.messages
+        if message.is_active and message.user_id == user_id and message.agent == agent
+    ]
+
+    if len(agent_messages) == 0:
```
Comment (tianjing-li) [inline src/backend/services/chat.py:281]: prefer: `if not agent_messages`

**Review by tianjing-li — APPROVED** [review]
Very solid PR. Only have one minor comment.

File: src/backend/services/chat.py (hunk 1 above)
Comment (danylo-boiko) [reply src/backend/services/chat.py]: Fixed, thanks!

**Review by tianjing-li — APPROVED** [review]

**Comment by tianjing-li** [comment]
@danylo-boiko merging, thanks!

**Pull request merged**